docs/specs/saml.md

```diff
@@ -0,0 +1,701 @@
+        - `pem`: Required. The X.509 cert in pem format.
+
+## <a id="7"></a> Metadata
+
+Metadata about the IdP will be provided through a metadata endpoint.
+
+The content of the metadata will be as mentioned in [SAMLMetadata](https://docs.oasis-open.org/security/saml/v2.0/saml-metadata-2.0-os.pdf). Including the following:
+
+- Supported bindings
+- Web Browser SSO endpoint
+- Single Logout endpoint
+- Signing algorithms and certs
+
+## <a id="8"></a> Debug Logging
+
+When `debug_mode_enabled` is `true`. Addtional non-blocking events will be created:
+
+- `saml.debug.sso.authn_reqeust_received`
+  - Triggered whenever a `<AuthnRequest>` received.
+- `saml.debug.sso.response_issued`
+  - Triggered whenever a `<Response>` issued.
+- `saml.debug.slo.logout_request_received`
+  - Triggered whenever a `<LogoutRequest>` received.
+- `saml.debug.slo.logout_request_issued`
+  - Triggered whenever a `<LogoutRequest>` issued.
+- `saml.debug.slo.logout_response_received`
+  - Triggered whenever a `<LogoutResponse>` received.
+- `saml.debug.slo.logout_response_issued`
+  - Triggered whenever a `<LogoutResponse>` issued.
+
+All of them will be saved as audit logs.
+
+The payload of the above events includes:
+
+- The request body.
+- The response body and the redirect URI if applicable.
+- The error on handling the request, if any.
+
+### <a id="8_1"></a> Configs
+
+```yaml
+saml:
+  debug_mode_enabled: true
+```
+
+## <a id="9"></a> Profiles not supported at the moment
+
+- Enhanced Client or Proxy (ECP) Profile
+  - No plan to support ECP clients.
+- Identity Provider Discovery Profile.
+  - No plan to support multiple IdP.
+- Name Identifier Management Profile
+  - This profile could be useful in the following case: Assume the `NameIDFormat` is set to `urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress`. And an admin changed the email address of the alice using admin API. In such case, Authgear as an IdP should notify SPs which holds an Name Ideentifier of alice using this profile.
+- Name Identifier Mapping Profile
+  - Not useful until we implemented the `urn:oasis:names:tc:SAML:2.0:nameid-format:persistent` NameIDFormat.
+  - Currently, the Name Identifiers issued by Authgear is always consistent accross different service providers.
+- Assertion query/request profile
+  - For querying existing assertions. Not implemented because all assertions will be provided in the authentication response.
+
+## <a id="10"></a> An example with all configurable options
+
+```yaml
+saml:
+  debug_mode_enabled: true
+  signing:
+    key_id: key01
+    signature_method: RSAwithSHA256
+    digest_method: SHA256
+  service_providers:
+    - id: ENTITY_ID
+      audience: https://example.com
```

Comment:
- [ ] `recipient` should also be configurable, as it might be different from acs_url if the sp is behind a proxy gateway.